Allow origin to be persisted if it is exempt from ITP deletion

szewai opened the pull request:
#### fb634d8ebf1e903515286602488a23367c6a1e61
<pre>
Allow origin to be persisted if it is exempt from ITP deletion
<a href="https://bugs.webkit.org/show_bug.cgi?id=256817">https://bugs.webkit.org/show_bug.cgi?id=256817</a>
rdar://109379695

Reviewed by Youenn Fablet.

According to spec (<a href="https://storage.spec.whatwg.org/#introduction)">https://storage.spec.whatwg.org/#introduction)</a>, an origin can request to be persistent with
StorageManager.persist(), which means its data will be exempt from eviction on storage pressure. This patch implements a
rule to decide whether an origin can be persistent.

Test: WKWebsiteDataStoreConfiguration.TotalQuotaRatioWithPersistedDomain

* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::ResourceLoadStatisticsStore::shouldExemptFromWebsiteDataDeletion const):
(WebKit::ResourceLoadStatisticsStore::domainsExemptFromWebsiteDataDeletion const):
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::registrableDomainsExemptFromWebsiteDataDeletion):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::registrableDomainsExemptFromWebsiteDataDeletion):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::close):
(WebKit::NetworkStorageManager::fetchRegistrableDomainsForPersist):
(WebKit::NetworkStorageManager::didFetchRegistrableDomainsForPersist):
(WebKit::NetworkStorageManager::persistOrigin):
(WebKit::NetworkStorageManager::persist):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::htmlStringForTotalQuotaRatioTest):
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/264220@main">https://commits.webkit.org/264220@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e0581d49c296e3219f16207a76aa1dcdbe725fb0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7049 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7293 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7472 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8665 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7283 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7059 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8555 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7225 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10187 "100 failures") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7176 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7838 "2 new passes 4 flakes 1 failures") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6501 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8768 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/5201 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6410 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14177 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/6864 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6505 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/9376 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6988 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5718 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6353 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1671 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10549 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6735 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -624,8 +664,15 @@ void NetworkStorageManager::persist(const WebCore::ClientOrigin& origin, Complet
+
+    RunLoop::main().dispatch([weakThis = ThreadSafeWeakPtr { *this }]() mutable {
+        if (auto strongThis = weakThis.get())
+            strongThis->fetchRegistrableDomainsForPersist();
```
The issue is that the registrable domains might be different between the time persist is called by the JS app (at that point, it could be exempt from ITP deletion) and when the storage pressure happens (say a few days later) at which point this domain might no longer be exempt from ITP deletion.
In that case, what should happen? Should we delete the data, defeating persistency, or keep the data, defeating ITP?

Also, this API would expose whether a website is exempt from ITP deletion or not. So far, I do not think we expose that bit of information to web pages, that could be used as a fingerprint information.

Another approach could be:
- Add a callback to application to ask whether a domain should be persisted or not. Default implementation might be yes. If doing so, we might need to surface to the application the persisting boolean flag.
- When storage pressure happens, clear everything possible except for storage marked as persisting.
- If this is not enough, expose a callback to notify the application that storage is too big and that it should take some actions (including clearing persisting data). I am not sure what the default implementation should be if the callback is not implemented, maybe delete persisting data by last time domain was visited?

szewai replied:
Discussed on slack. (Domains exempt from ITP deletion are configured by WebKit clients, which should be fixed; they are not equal to domains that ITP decides to keep data, which can change during runtime).

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -614,6 +616,44 @@ void NetworkStorageManager::persisted(const WebCore::ClientOrigin& origin, Compl
+
+        for (auto&& [origin, completionHandler] : std::exchange(m_persistCompletionHandlers, { })) {
+            if (!domains.contains(origin.clientRegistrableDomain()))
+                completionHandler(false);
```
In theory, we should delete the file if it exists.

szewai replied:
Okay, will update.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -624,8 +664,15 @@ void NetworkStorageManager::persist(const WebCore::ClientOrigin& origin, Complet
-    return completionHandler(false);
+    bool shouldFetchRegistrableDomains = m_persistCompletionHandlers.isEmpty();
+    m_persistCompletionHandlers.append({ origin, WTFMove(completionHandler) });
+    if (!shouldFetchRegistrableDomains)
```
We probably need to fetch these domains only once since they are static info.
This will make it clearer that these domains are exempt from ITP classification, not that they are classified as non trackers.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
```diff
@@ -266,6 +269,9 @@ class NetworkStorageManager final : public IPC::WorkQueueMessageReceiver {
     HashMap<IPC::Connection::UniqueID, HashSet<String>> m_temporaryBlobPathsByConnection WTF_GUARDED_BY_CAPABILITY(workQueue());
+    using OriginPersistCompletionHandler = std::pair<WebCore::ClientOrigin, CompletionHandler<void(bool)>>;
+    Vector<OriginPersistCompletionHandler> m_persistCompletionHandlers WTF_GUARDED_BY_CAPABILITY(workQueue());
+    std::optional<HashSet<WebCore::RegistrableDomain>> m_domainsExemptFromEviction;
```
WTF_GUARDED_BY_CAPABILITY as well?

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -614,6 +617,55 @@ void NetworkStorageManager::persisted(const WebCore::ClientOrigin& origin, Compl
+    });
+}
+
+void NetworkStorageManager::persistOrigin(const WebCore::ClientOrigin& origin, CompletionHandler<void(bool)>&& completionHandler)
```
It would be better to write it as:
`bool persistOrigin(const WebCore::ClientOrigin&)`

szewai committed:
Allow origin to be persisted if it is exempt from ITP deletion
https://bugs.webkit.org/show_bug.cgi?id=256817
rdar://109379695

Reviewed by Youenn Fablet.

According to spec (https://storage.spec.whatwg.org/#introduction), an origin can request to be persistent with
StorageManager.persist(), which means its data will be exempt from eviction on storage pressure. This patch implements a
rule to decide whether an origin can be persistent.

Test: WKWebsiteDataStoreConfiguration.TotalQuotaRatioWithPersistedDomain

* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::ResourceLoadStatisticsStore::shouldExemptFromWebsiteDataDeletion const):
(WebKit::ResourceLoadStatisticsStore::domainsExemptFromWebsiteDataDeletion const):
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::registrableDomainsExemptFromWebsiteDataDeletion):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::registrableDomainsExemptFromWebsiteDataDeletion):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::close):
(WebKit::NetworkStorageManager::fetchRegistrableDomainsForPersist):
(WebKit::NetworkStorageManager::didFetchRegistrableDomainsForPersist):
(WebKit::NetworkStorageManager::persistOrigin):
(WebKit::NetworkStorageManager::persist):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::htmlStringForTotalQuotaRatioTest):
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/264220@main
